lib/puppetserver/ca/sign_action.rb

```diff
@@ -0,0 +1,252 @@
+            parsed['help'] = true
+          end
+          opts.on('--version', 'Output the version') do |v|
+            parsed['version'] = true
+          end
+          opts.on('--all', 'Operate on all certnames') do |a|
+            parsed['all'] = true
+          end
+        end
+      end
+
+      def initialize(logger)
+        @logger = logger
+      end
+
+      def run(input)
+        if input['all']
+          requestedCertnames = get_all_pending_certs
+          if requestedCertnames.nil?
+            return 1
+          end
+        else
+          requestedCertnames = input['certname']
+        end
+
+        exit_code = sign_requested_certs(requestedCertnames)
+        return exit_code
+      end
+
+      def logger_output_put(result,certname)
+        if result.is_a? Net::HTTPNoContent
+          @logger.inform "Signed Certificate for #{certname}"
+          return 0
+        elsif result.is_a? Net::HTTPNotFound
+          @logger.err 'Error'
+          @logger.err "    Could not find certificate for #{certname}"
+          return 1
+        else
+          @logger.err 'Error - gross debug output:'
+          @logger.err "    #{result.inspect}"
+          return 1
+        end
+      end
+
+      def get_certificate_statuses
+        uri = URI('https://localhost:8140/puppet-ca/v1/certificate_statuses/:any_key?environment=production')
+        Net::HTTP.start(uri.host, uri.port, use_ssl: true, cert_store: store) do |connection|
+          request = Net::HTTP::Get.new uri
+          connection.request(request)
+        end
+      end
+
+      def sign_cert(certname)
+        uri = URI('https://localhost:8140/puppet-ca/v1/certificate_status/' + certname + '?environment=production')
+        Net::HTTP.start(uri.host, uri.port, use_ssl: true, cert_store: store ) do |connection|
+          request = Net::HTTP::Put.new(uri, HEADERS)
+          request.body = BODY
+          connection.request(request)
+        end
+      end
+
+      def get_all_certs
+        result = get_certificate_statuses
+        if !result.is_a? Net::HTTPOK
+            @logger.err 'Error - gross debug output:'
+            @logger.err "    #{result.inspect}"
+            return nil
+        end
+        return result
+      end
+
+      def select_pending_certs(get_result)
+        requestedCertnames = JSON.parse(get_result.body).select{|e| e["state"] == "requested"}.map{|e| e["name"]}
+
+        if requestedCertnames.empty?
+          @logger.err 'Error'
+          @logger.err "No waiting certificate requests to sign"
+          return nil
+        end
+
+        return requestedCertnames
+      end
+
+      def get_all_pending_certs
+        result = get_all_certs
+        if result
+          select_pending_certs(result)
+        end
+      end
+
+      def sign_requested_certs(certnames)
+        exit_code = 0
+        certnames.each do |certname|
+          result = sign_cert(certname)
+
+          if result.is_a? Net::HTTPNoContent
```

Comment:
@pcarlisle would you put the result handling here in its own method? I'm wondering how thinly different folks would slice these abstractions.